server/src/main/java/com/saralein/server/request/RequestParser.java

```diff
@@ -23,50 +32,48 @@ public Request parse(String rawRequest) throws Exception {
         }
     }
 
-    private Request buildRequest(String[] requestLine, List<String> request) {
-        return new Request.Builder()
-                .method(getMethod(requestLine))
-                .uri(getUri(requestLine))
-                .body(parseBody(request))
-                .addHeaders(parseHeaders(request))
-                .build();
-    }
-
-    private String getMethod(String[] requestLine) {
-        return requestLine[0];
+    private Map<String, String> parseHeaders(List<String> request) {
+        return selectHeaderLines(request)
+                .stream()
+                .map(this::splitNameValue)
+                .collect(Collectors.toMap(line -> line.get(0), line -> line.get(1)));
     }
 
-    private String getUri(String[] requestLine) {
-        return requestLine[1];
+    private List<String> selectHeaderLines(List<String> request) {
+        return requestHasBody(request)
+                ? new ArrayList<>(request.subList(1, request.size() - 1))
+                : new ArrayList<>(request.subList(1, request.size()));
     }
 
-    private String parseBody(List<String> request) {
-        String lastLine = request.get(request.size() - 1);
-
-        return startsWithBodyTag(lastLine) ? getBodyValue(lastLine) : "";
+    private boolean requestHasBody(List<String> request) {
+        return request.stream()
+                .anyMatch(line -> line.startsWith("Content-Length"));
```

Comment:
Will this work if the request has `Content-Length: 0`?

Reply:
Doh.  Added a fix and test for that.